src/AccessTokenProvider.js

```diff
@@ -0,0 +1,127 @@
+import {
+  isResponseUnauthorized,
+} from './services/http';
+
+/**
+ * Provides a way for renewing access token with correct renew token. It will automatically
+ * dispatch a call to server with request provided via config. It also ensures that
+ * renew token is fetched only once no matter how many requests are trying to get
+ * a renewed version of access token at the moment. All subsequent requests will be chained
+ * to renewing fetch promise and resolved once the response is received.
+ */
+export default class AccessTokenProvider {
+  constructor(fetch, config) {
+    this.fetch = fetch;
+
+    this.config = config;
+    this.renewAccessTokenPromise = null;
+    this.tokens = {
+      refreshToken: null,
+      accessToken: null,
+    };
+
+    this.renew = this.renew.bind(this);
+    this.authorize = this.authorize.bind(this);
+    this.getAuthorization = this.getAuthorization.bind(this);
+    this.clear = this.clear.bind(this);
+
+    this.isAuthorized = this.isAuthorized.bind(this);
+    this.resolveAccessToken = this.resolveAccessToken.bind(this);
+    this.fetchAccessToken = this.fetchAccessToken.bind(this);
+    this.handleFetchAccessTokenResponse = this.handleFetchAccessTokenResponse.bind(this);
+    this.handleAccessToken = this.handleAccessToken.bind(this);
+    this.handleError = this.handleError.bind(this);
+  }
+
+  /**
+   * Renews current access token with provided refresh token
+   */
+  renew() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      console.log('Please authorize provider before renewing or check shouldIntercept config.');
+      return Promise.resolve();
+    }
+
+    // if we are not running token promise, start it
+    if (!this.renewAccessTokenPromise) {
+      this.renewAccessTokenPromise = new Promise(this.resolveAccessToken);
+    }
+
+    // otherwise just return existing promise
+    return this.renewAccessTokenPromise;
+  }
+
+  /**
+   * Authorizes intercept library with given renew token
+   * @param refreshToken
+   * @param accessToken
+   */
+  authorize(refreshToken, accessToken) {
```

Comment:
I want to allow passing both refresh token and access token for two reasons:
1. perhaps you already have both stored externally and both are valid (why forcing a renew?)
2. RFC spec allows authorization response to return both https://tools.ietf.org/html/rfc6749#page-10